tests/test_fields.py

```diff
@@ -873,6 +873,11 @@ def __init__(self, a):
         self.assertEqual(len(form.a.entries), 3)
         self.assertEqual(form.a.data, data)
 
+    def test_errors(self):
+        F = make_form(a=FieldList(self.t))
+        form = F(DummyPostData({'a-0': ['a'], 'a-1': ''}))
+        assert not form.validate()
+        self.assertEqual(form.a.errors, [None, ['This field is required.']])
```

Comment:
This doesn't solve the actual issue of enumerating which field within the field enclosure, i.e - 'a-0-user_name'. I thought that was the whole purpose of the PR?

Reply:
I don't understand what you mean. I'll explain it another way:

Currently, `FormList.errors` loses information. Take the output `[<error>, <error>]`: this could have been produced by many inputs: `[<invalid>, <invalid>]`, `[<valid>, <invalid>, <invalid>]`, `[<invalid>, <invalid>, <valid>]`, etc.

Or to put it another way, the indices in the `errors` don't necessarily match the indices in `data`. By inserting a dummy value (`None` or `[]`), the indices can be made to match up, and information isn't lost anymore.

Reply:
You're not actually doing anything though. The errors themselves are still identical from this project vs. your fork. 
Besides that, if I again think what you are saying is correct than you want to have an error be reflected like this:

```python
class IMForm(Form):
    protocol = SelectField(choices=[('aim', 'AIM'), ('msn', 'MSN')])
    username = StringField()

class ContactForm(Form):
    first_name  = StringField()
    last_name   = StringField()
    im_accounts = FieldList(FormField(IMForm))

# for brevity assume data has 2 entries for IMForm
# would look like this 'im_accounts-0-protocol' etc..
form = ContactForm(data)
form.validate()  # is false for now
form.errors  # should be a dictionary such as {'im_accounts-0-protocol': ['Not a valid choice']}

```

Thats what I thought you're trying to say. If thats what you're trying to say then I agree it should help and enumerate which field is exactly missing/incorrect. Having said that this PR again is not doing as such. Theres virtually no change in the `form.errors` and just does more work for no clear reason. Your test also needs to be more clear in what its doing and what should be the exact structure of your modified form.errors dictionary.

Reply:
I understand: you want `form.errors` to have the same structure as `formdata`, I want it to have the same structure as `data`: `{ 'first_name': '...', 'im_accounts': [{ 'protocol': '...' }, ...] }`.

Currently, `Form.errors`/`Field.errors` are `data`-structured. This PR intended to correct the bug in `FieldList.errors` under that assumption.

Personally, I prefer `data`-structured approach. `formdata` structure, IMO, should only be used on the input side as a convenient "parser" of HTML form data (since HTML doesn't have proper support for nested data).

Reply:
I don't have a preference for the structure, I'm only mentioning one way of correctly enumerating the associated field that got the error. 

Also the current errors are just the way you want them to be. When using FieldList the error will be structured as such:
```python
{
    'first_name': ['This field is required.'],
    'im_accounts': [{'protocol': ['This field is required.'],
    'username': ['This field is required.']}]
}

```
As you can see there is no enumeration of the fields within the errors which is what I am pointing out. I thought thats what you meant. If it isn't, please point out what it is that you wanted if not that?

Reply:
```python
listdata = ['a', '']
form = F(data = { 'a': ['a', ''] })
for i, (item, item_errors) in enumerate(zip(listdata, form.a.errors)):
    print("Errors for item {} (value = {!r}): {!r}".format(i, item, item_errors))
```

Re: "The errors themselves are still identical from this project vs. your fork." The errors are identical, but the whole `FieldList.errors` object is not: it now contains "padding" objects so that errors at index `i` correspond to the data at index `i`.

Reply:
Could you please provide the errors dictionary for your fork  and the current dictionary for wtforms ? If you can use the example `class ContactForm`. I've got your code loaded up and running locally and maybe I'm missing something.

Reply:
Either the `form.errors` or the `form.im_accounts.errors` that is.

Reply:
```python
form = ContactForm(data = {
	'first_name': "Test",
	'last_name': "Test",
	'im_accounts': [{ 'protocol': 'aim' }, {}]
})
assert not form.validate()
print(form.errors)

# Before:
{'im_accounts': [{'protocol': ['Not a valid choice']}]}
# After:
{'im_accounts': [None, {'protocol': ['Not a valid choice']}]}
```

Reply:
On a clean install Ubuntu 17.10 with a copy of your fork.
```python

class IMForm(Form):
    protocol = SelectField(choices=[('aim', 'AIM'), ('msn', 'MSN')])
    username = StringField()

class ContactForm(Form):
    first_name  = StringField()
    last_name   = StringField()
    im_accounts = FieldList(FormField(IMForm))
data = {}
data['first_name'] = 'test_first_name'
data['last_name'] = 'test_last_name'
data['im_accounts'] = [{'protocol':'gchat', 'username': 'test_username'}]
form = ContactForm(data=data)
form.validate()
form.errors # >> {'im_accounts': [{'protocol': ['Not a valid choice']}]}
form.im_accounts.errors # >> [{'protocol': ['Not a valid choice']}]

```

Different server same Ubuntu 17.10 but this is straight from the `pip install -U WTForms`

```python
class IMForm(Form):
    protocol = SelectField(choices=[('aim', 'AIM'), ('msn', 'MSN')])
    username = StringField()

class ContactForm(Form):
    first_name  = StringField()
    last_name   = StringField()
    im_accounts = FieldList(FormField(IMForm))
data = {}
data['first_name'] = 'test_first_name'
data['last_name'] = 'test_last_name'
data['im_accounts'] = [{'protocol':'gchat', 'username': 'test_username'}]
form = ContactForm(data=data)
form.validate()
form.errors # >> {'im_accounts': [{'protocol': ['Not a valid choice']}]}
form.im_accounts.errors # >> [{'protocol': ['Not a valid choice']}]
```

As you can see they are identical.

Reply:
@whb07 The result would be the same in that case, since `data['im_accounts']` only contains _one_ item.

Reply:
Yes I see what you're saying now. But that doesn't do anything to help the user figure out which field in the data is invalid.
`{'im_accounts': [{'protocol': ['Not a valid choice']}, None]}` thats the dictionary that comes from your changes when passed two `im_accounts`. 

So we are back to guessing what exactly is wrong with the data except that instead of looking at say N fields, I'm still looking at N fields because its not clear which particular data point is invalid. Except now I have a (N - correct-fields) worth of `None` objects to look at.